### ISSUE TITLE
EngineCrash [master f83afe1] at RenderUnitDestroyed on synthetic test

### ISSUE THREAD

**Issue by saurtron** [beyond-all-reason/spring#1810]
Found a case where the engine crashes, probably not very worrying since it just happens on a synthetic test.

at: `rts/Lua/LuaHandle.cpp:1925  CLuaHandle::RenderUnitDestroyed(CUnit const*)`

Seems to be because of cleaning the map just after the unit selfd, likely between when the unit is destroyed and the callin runs. I guess the unit is expected to still be there but the test map clearing already deleted it itself.

### Reproducing

- Take [test_crash.lua](https://github.com/saurtron/Beyond-All-Reason/blob/test-crash/luaui/Widgets/Tests/crash/test_crash.lua) and place under luaui/Widgets/Tests/crash/
- Start skirmish in neverend mode
- Call /runtests test_crash

After a few runs it usually crashes, but could depend on the specific system.

**Comment by sprunk** [comment]
> After a few runs it usually crashes, but could depend on the specific system.

My guess the clear has to happen on a frame the unit is undergoing the death animation (which even units without animation do, they exist in the "dead" state for like 2 frames or so). The lack of 100% reproducibility would be because self-d only ticks on slowupdate so will randomly take 0-16 more frames than expected, see #1713. You can poll whether a unit is in the dying state via `Spring.GetUnitIsDead` which returns `true` for units in the process of dying and `nil` when the unit is completely gone.

**Comment by saurtron** [comment]
> My guess the clear has to happen on a frame the unit is undergoing the death animation (which even units without animation do, they exist in the "dead" state for like 2 frames or so). The lack of 100% reproducibility would be because self-d only ticks on slowupdate so will randomly take 0-16 more frames than expected, see #1713. You can poll whether a unit is in the dying state via `Spring.GetUnitIsDead` which returns `true` for units in the process of dying and `nil` when the unit is completely gone.

You nailed it, by making the test wait unitl `Spring.GetUnitIsDead` I can now make it crash every time on the first run.

Updated [test_crash.lua](https://github.com/saurtron/Beyond-All-Reason/blob/test-crash/luaui/Widgets/Tests/crash/test_crash.lua) for that.

Now just need to check the code and figure out a solution, I guess not running the callin if the unit is destroyed, although I'm thinking this might be a problem in other situations too? so maybe worth it to dig a bit deeper than that.

**Comment by sprunk** [comment]
> although I'm thinking this might be a problem in other situations too?

Off the top of my head I'd expect a crash in `wupget:UnitDamaged` (or earlier) if you remove the unit within `wupget:UnitPreDamaged`, or in `wupget:DrawUnitReflection` if you remove it in `wupget:DrawUnit`. But then I'd also expect a crash/hang if you do `while true do ; end`. I'm not sure to what extent we can prevent game devs from causing crashes, but most existing ways are going to be things that sound prima facie stupid, like the above, so I don't think we need to worry about those others too much until somebody shows up with a use case. Simply removing all units in a loop sounds like a reasonable thing to want though so it would be good to make that working.

**Comment by saurtron** [comment]
> Simply removing all units in a loop sounds like a reasonable thing to want though so it would be good to make that working.

Yes, this happened to me while working on test files, so it's kind of a real use case since the tests destroy all units all the time after running every test.

Not a blocker since only affects tests/devs, but still.
